File: src/main/java/org/oewntk/model/Groupings.java
/*
 * Copyright (c) 2021. Bernard Bou.
 */

package org.oewntk.model;

import java.util.*;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

import static java.util.stream.Collectors.*;

/**
 * Generic grouping factory
 */
public class Groupings
{
	// group by
	public static <K, V> Map<K, Collection<V>> groupBy(final Collection<V> things, final Function<V, K> groupingFunction)
	{
		return groupBy(things.stream(), groupingFunction);
	}

	public static <K, V> Map<K, Collection<V>> groupBy(final Stream<V> stream, final Function<V, K> groupingFunction)
	{
		return stream //
				.collect(groupingBy(groupingFunction, toCollection(HashSet::new)));
	}

	// counts

	public static <K, V> Map<K, Long> countsBy(final  Stream<V> stream, final Function<V, K> groupingFunction)
	{
		return stream //
				.collect(groupingBy(groupingFunction, TreeMap::new, counting()));
	}

	public static <K, V> Map<K, Long> multipleCountsBy(final  Stream<V> stream, final Function<V, K> groupingFunction)
	{
		return stream //
				.collect(collectingAndThen( //
						groupingBy(groupingFunction, TreeMap::new, counting()), //
						m -> {
							m.values().removeIf(v -> v <= 1L);
							return m;
						}));
	}

	// group by having

	public static <K, V> Map<K, List<V>> groupByHaving(final  Stream<V> stream, final Function<V, K> groupingFunction, final Predicate<List<V>> predicate)
	{
		return stream //
				.collect(collectingAndThen( //
						groupingBy(groupingFunction, TreeMap::new, toList()), //
						map -> {
							map.values().removeIf(predicate);
							return map;
						}));
	}

	public static <K, V> Map<K, List<V>> groupByHavingMultipleCount(final  Stream<V> stream, final Function<V, K> groupingFunction)
	{
		return groupByHaving(stream, groupingFunction, vals -> vals.size() <= 1L);
	}
}

File: src/main/java/org/oewntk/model/Groupings.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

/**
 * Generic grouping factory
 */
object Groupings {

    // group by having

    /**
     * Group elements having
     *
     * @param collection       collection of elements
     * @param groupingFunction map element to key
     * @param predicate        having clause : boolean function of a set of values
     * @param K                grouping key
     * @param V                type of elements
     * @return list of elements grouped and mapped by key
     */
    private fun <K : Comparable<K>, V> groupByHaving(
        collection: Collection<V>,
        groupingFunction: (V) -> K,
        predicate: (Set<V>) -> Boolean,
    ): Map<K, Set<V>> {
        return collection
            .groupBy(groupingFunction) // mapOf(K, listOf(V1, V2, ..))
            .mapValues { it.value.toSet() } // mapOf(K, setOf(V1, V2, ..))
            .toList() // listOf(K to setOf(V1, V2, ..))
            .filterNot { predicate.invoke(it.second) } // filtered listOf(K to setOf(V1, V2, ..))
            .toMap() // mapOf(K to setOf(V1, V2, ..))
            .toSortedMap(naturalOrder()) // sortedMapOf(K to setOf(V1, V2, ..))
    }

    /**
     * Group elements having group count &gt; 1
     *
     * @param collection       collection of elements
     * @param groupingFunction map element to key
     * @param K                grouping key
     * @param V                type of elements
     * @return list of elements grouped and mapped by key
     */
    fun <K : Comparable<K>, V> groupByHavingMultipleCount(collection: Collection<V>, groupingFunction: (V) -> K): Map<K, Set<V>> {
        return groupByHaving(collection, groupingFunction) { values -> values.size <= 1L }
    }

    // counts

    /**
     * Group elements by key and yield count of each group
     *
     * @param collection       collection of elements
     * @param groupingFunction map element to key
     * @param K                type of key
     * @param V                type of element
     * @return count for each key/group
     */
    fun <K : Comparable<K>, V> countsBy(collection: Collection<V>, groupingFunction: (V) -> K): Map<K, Long> {
        return collection
            .groupBy(groupingFunction) // mapOf(K, listOf(V1, V2, ..))
            .toSortedMap(naturalOrder()) // sortedMapOf(K, listOf(V1, V2, ..))
            .mapValues { it.value.toSet().size.toLong() } // sortedMapOf(K, sizeOf(listOf(V1, V2, ..)))
    }

    /**
     * Group elements by key and yield count of each group, retain only groups whose count &gt; 1
     *
     * @param collection       collection of elements
     * @param groupingFunction map element to key
     * @param K                type of key
     * @param V                type of element
     * @return count for each key/group
     */
    fun <K : Comparable<K>, V> multipleCountsBy(collection: Collection<V>, groupingFunction: (V) -> K): Map<K, Long> {
        return collection
            .groupBy(groupingFunction) // mapOf(K, listOf(V1, V2, ..))
            .mapValues { it.value.toSet().size.toLong() } // mapOf(K, setOf(V1, V2, ..).size)
            .toList() // listOf(K to size)
            .filter { it.second > 1L } // filtered listOf(K to size)
            .toMap() // mapOf(K to size)
            .toSortedMap(naturalOrder()) // sortedMapOf(K to size)
    }
}
